gensim/models/poincare.py

```diff
@@ -0,0 +1,637 @@
+            # Move all updates to the same node to the last such update, zeroing all the others
+            vector_updates[positions[-1]] = vector_updates[positions].sum(axis=0)
+            vector_updates[positions[:-1]] = 0
+
+    def update_vectors_batch(self, batch):
+        """
+        Updates vectors for nodes in the given batch.
+
+        Parameters
+        ----------
+        batch : PoincareBatch instance
+            batch containing computed gradients and node indices of the batch for which updates are to be done.
+
+        Notes
+        -----
+        Mutates the `syn0` array.
+        """
+        grad_u, grad_v = batch.gradients_u, batch.gradients_v
+        indices_u, indices_v = batch.indices_u, batch.indices_v
+        batch_size = len(indices_u)
+
+        u_updates = (self.alpha * (batch.alpha ** 2) / 4 * grad_u).T
+        self.handle_duplicates(u_updates, indices_u)
+
+        self.wv.syn0[indices_u] -= u_updates
+        self.wv.syn0[indices_u] = self.clip_vectors(self.wv.syn0[indices_u], self.epsilon)
+
+        v_updates = self.alpha * (batch.beta ** 2)[:, np.newaxis] / 4 * grad_v
+        v_updates = v_updates.swapaxes(1, 2).swapaxes(0, 1)
+        v_updates = v_updates.reshape(((1 + self.negative) * batch_size, self.size))
+        self.handle_duplicates(v_updates, indices_v)
+
+        self.wv.syn0[indices_v] -= v_updates
+        self.wv.syn0[indices_v] = self.clip_vectors(self.wv.syn0[indices_v], self.epsilon)
+
+    def train(self):
+        """Trains Poincare embeddings using loaded data and model parameters."""
+        if self.burn_in > 0:
+            self.alpha = self.burn_in_alpha
+            self.train_batchwise(epochs=self.burn_in)
+        self.alpha = self.train_alpha
+        self.train_batchwise()
+
+    def train_batchwise(self, epochs=None, batch_size=10, print_every=1000):
+        """
+        Trains Poincare embeddings using specified parameters.
+
+        Parameters
+        ----------
+        epochs : int or None, optional
+            Number of epochs after which training ends, if `None`, runs for `self.iter` epochs, default `None`.
+        batch_size : int, optional
+            Number of examples to train on in a single batch, defaults to 10.
+        print_every : int or None, optional
+            Prints progress and average loss after every `print_every` batches, defaults to 1000.
+        """
+        if self.workers > 1:
+            raise NotImplementedError("Multi-threaded version not implemented yet")
+        if epochs is None:
+            epochs = self.iter
+        for epoch in range(1, epochs + 1):
+            indices = list(range(len(self.all_relations)))
+            self.np_random.shuffle(indices)
+            avg_loss = 0.0
+            last_time = time.time()
+            for batch_num, i in enumerate(range(0, len(indices), batch_size), start=1):
+                print_check = not (batch_num % print_every)
```

Comment:
I prefer `should_do_x` over `x_check` for boolean flags of this kind. "if should print" reads more naturally than "if print check".